### PR TITLE
Added some support for the publish() method to return false if the broker rejects a message

### PR DESCRIPTION
**Pull request by petekelly**
One of the features I need for my project is to know for sure whether a message has been accepted (and queued) by the broker.

Previously we used to use the phplib-amqp library directly but we have now moved to Laravel and are liking the ease of use of laravel-amqp.

I've added the changed we have made so that we are notified if the broker cannot queue a message. It requires putting the publisher into confirm_select mode, and also waiting for a broker response before closing the connection,  so it is not for every circumstance!

Please feel free to merge in if this is a feature that would be useful - or if you would prefer it to be done a different way, or any changes made please let me know.

### CODE REVIEW

File: src/Publisher.php [hunk 1]
```diff
@@ -7,14 +7,44 @@
...
+    */
+    private $publish_result = null;
 
     /**
      * @param string  $routing
      * @param Message $message
      * @throws Exception\Configuration
      */
-    public function publish($routing, $message)
+    public function publish($routing, $message, $mandatory=false)
```
Comment (stevenklar) [inline src/Publisher.php]: Please add a space between variable and value
```suggestion
    public function publish($routing, $message, $mandatory = false)
```

File: src/Publisher.php [hunk 2]
```diff
@@ -7,14 +7,44 @@
...
+            $this->getChannel()->confirm_select();
+            $this->getChannel()->set_nack_handler([$this, 'nack']);
+            $this->getChannel()->set_return_listener([$this, 'return']);
+        }
+
+        $this->getChannel()->basic_publish($message, $this->getProperty('exchange'), $routing, $mandatory);
+        $mandatory == true && $this->getChannel()->wait_for_pending_acks_returns(30);
+    }
+
+    /*
```
Comment (stevenklar) [inline src/Publisher.php]: Please use the common phpDoc format for methods. http://docs.phpdoc.org/references/phpdoc/tags/method.html

File: src/Publisher.php [hunk 3]
```diff
@@ -7,14 +7,44 @@
...
      * @param Message $message
      * @throws Exception\Configuration
      */
-    public function publish($routing, $message)
+    public function publish($routing, $message, $mandatory=false)
     {
-        $this->getChannel()->basic_publish($message, $this->getProperty('exchange'), $routing);
+	$this->publish_result = true;
+
+	if($mandatory == true) {
```
Comment (stevenklar) [inline src/Publisher.php]:
```suggestion
	if ($mandatory === true) {
```
Check format

File: src/Publisher.php [hunk 4]
```diff
@@ -7,14 +7,44 @@
...
+            $this->getChannel()->confirm_select();
+            $this->getChannel()->set_nack_handler([$this, 'nack']);
```
Comment (stevenklar) [inline src/Publisher.php]: If we add the nack handler we should also add the ack handler, what do you think?

Also I would like to make this more configureable.
If I define a ack/nack/return handler it should be set for the channel.

File: src/Publisher.php [hunk 5]
```diff
@@ -7,14 +7,44 @@
...
+        }
+
+        $this->getChannel()->basic_publish($message, $this->getProperty('exchange'), $routing, $mandatory);
+        $mandatory == true && $this->getChannel()->wait_for_pending_acks_returns(30);
+    }
+
+    /*
+    * @param Object Message object
+    */
+    public function nack($msg) {
```
Comment (stevenklar) [inline src/Publisher.php]:
```suggestion
    public function nack(AMQPMessage $message) {
```
We can be type safe here as the handler should always return a proper AMQPMessage. Right?

File: src/Publisher.php [hunk 6]
```diff
@@ -7,14 +7,44 @@
...
+    	$this->publish_result = false;
+    }
+
+    /*
```
Comment (stevenklar) [inline src/Publisher.php]: PhpDoc Format see above.

File: src/Publisher.php [hunk 7]
```diff
@@ -7,14 +7,44 @@
...
+	$this->publish_result = true;
+
+	if($mandatory == true) {
+            $this->getChannel()->confirm_select();
+            $this->getChannel()->set_nack_handler([$this, 'nack']);
+            $this->getChannel()->set_return_listener([$this, 'return']);
+        }
+
+        $this->getChannel()->basic_publish($message, $this->getProperty('exchange'), $routing, $mandatory);
+        $mandatory == true && $this->getChannel()->wait_for_pending_acks_returns(30);
```
Comment (stevenklar) [inline src/Publisher.php]: Where does the magic number 30 come from?

File: src/Publisher.php (hunk 5 above)
Comment (petekelly) [reply src/Publisher.php]: It actually looks like it is an array of AMQPMessages... :/

From AMQPChannel.php - $this->dispatch_to_handler($handler, array($message));

File: src/Publisher.php (hunk 7 above)
Comment (petekelly) [reply src/Publisher.php]: I chose 30 (seconds) arbitrarily. I can make it a definable value if you wish?

**Comment by petekelly** [comment]
> Thank you for this contribution.
> 
> I made some syntax notes and have some questions. Please take your time and review them carefully.
> Also what is the reason for the `$this->publish_result` to be set?

It's an async process so the object is holding the final value of the publish result. I defaulted it to true as the default behaviour of Rabbit/AMQP seems to be to fire and forger unless we are in the confirm mode.

File: src/Publisher.php (hunk 7 above)
Comment (stevenklar) [reply src/Publisher.php]: Yep it should be configurable please. But the 30 is a good default.

File: src/Publisher.php (hunk 6 above)
Comment (stevenklar) [reply src/Publisher.php]: Yes

File: src/Publisher.php (hunk 7 above)
Comment (petekelly) [reply src/Publisher.php]: I have made this configurable.

File: src/Publisher.php (hunk 6 above)
Comment (petekelly) [reply src/Publisher.php]: Fixed!

File: src/Publisher.php (hunk 5 above)
Comment (petekelly) [reply src/Publisher.php]: Updated the PHPDoc to reflect that this is an array.

**Comment by stevenklar** [comment]
I still need some time testing the changes and think about it, please stay tuned. Thank you for the changes so far.

**Comment by pedroboudoux** [comment]
Greetings,
I noticed that this feature has been without updates for a while (over a year), do we have any chance of having this into the project?

**Comment by bschmitt** [comment]
I hope I will have some more time soon, I might also have a look

**Comment by justasSendrauskas** [comment]
would be really good to get this in, please and thank you

**Comment by petekelly** [comment]
@bschmitt - I've just rebased this. Any chance it can be merged, I've been using it successfully for years now on production systems.

**Comment by bschmitt** [comment]
@petekelly thank you very much.

**Pull request merged**